app/webpacker/controllers/unsaved_changes_controller.js

```diff
@@ -0,0 +1,54 @@
+import { Controller } from "stimulus";
+
+// UnsavedChanges allows you to promp the user about unsaved changes when trying to leave the page
+//
+// Usage :
+// - with beforeunload event :
+//    <form data-controller="unsaved-changes" data-action="beforeunload@window->unsaved-changes#leavingPage" data-unsaved-changes-changed="true">
+//      <input data-action="change->unsaved-changes#formIsChanged">
+//    </form>
+//
+// - with turbolinks :
+//    <form data-controller="unsaved-changes" data-action="turbolinks:before-visit@window->unsaved-changes#leavingPage" data-unsaved-changes-changed="true">
+//      <input data-action="change->unsaved-changes#formIsChanged">
+//    </form>
+//
+// You can also combine the two actions
+//
+export default class extends Controller {
+  formIsChanged(event) {
+    this.setChanged("true");
```

Comment:
I'm still dreaming of a way of comparing the initial state to the current state to determine this value. It means that changing something back marks the form as clean. I think that there have been some attempts but nobody got it quite right. Or did I forget? My best approach would be take the parameters of the initial form and compare them to what would be submitted now. You don't have to do this now though. Probably best to finish the migration away from AngularJS first.

Reply:
I was thinking the same thing; surely this can be solved (and solved well) by a plugin. But I think this method is sufficient.

And great documentation!

Reply:
I was thinking: 

could it be something that is automatic like (in `connect()` method):
```js
this.element
      .querySelectorAll("input, select, textarea")
      .forEach((input) => {
        input.addEventListener("change", this.formIsChanged.bind(this));
      });
```
This avoid to use `data-action="change->unsaved-changes#formIsChanged"` for each form element. 
If we don't want to enabled that for each form elements, why not using `target` then?

It's not really important, but to me it's more easy to understand and much simpler. 

What do you think?

Reply:
Yeah I like the idea and it would make it easier to use. I did think about doing something similar over the week-end (you know shower thought ) :smile:  
So if I understand correctly we could do something like this in controller : 

```javascript
static targets = [ "formIsChanged ]

...

this.formIsChangedTargets
  .forEach((input) => {
    input.addEventListener("change", this.formIsChanged.bind(this));
  });
```
and in the html we would have :
```html
<input data-unsaved-changes-target="formIsChanged"  />
```
Maybe it would make it easier to understand, as you'd see in controller directly what triggers the update to `changed`.
I would be in favour of doing what I just described, it would probably set a better precedent that my current solution. 

What do people think ? @mkllnk @dacook @jibees  ?

Reply:
What's the advantage of using targets instead of the simple `connect` method?

I now found where it came up before:

- https://github.com/openfoodfoundation/openfoodnetwork/pull/9807

That PR proposes a custom controller but we can probably make our life easier with this library which was discussed at the time:

- https://github.com/kirillplatonov/dirty-form

Unfortunately, we were not clear in which direction to go and the contributors lost interest. But if we introduce this as the core team then future contributors can use that, too.

Reply:
@mkllnk 
> What's the advantage of using targets instead of the simple connect method?

What do you mean exactly by _simple connect method_? I'm not sure to understand: you were referencing my code?

@rioug @mkllnk 
Regarding https://github.com/kirillplatonov/dirty-form they use all elements of the form, without any target (nor action):
```
this.fields = [
      ...this.form.elements,
      ...this.form.querySelectorAll('trix-editor')
    ]
```
([source](https://github.com/kirillplatonov/dirty-form/blob/master/src/dirty-form.js#L6))

Reply:
> What's the advantage of using targets instead of the simple connect method?

Using the targets method (or the current action method) it would allow you to have some form element that would not make the form dirty. But then I am not sure if there is a use case for this really. It's probably something we can deal with if the need arises.

I'll give the library a go, to see if fits our needs.

Reply:
@jibees 
> What do you mean exactly by simple connect method? I'm not sure to understand: you were referencing my code?

Yes, I was referencing your code.

@rioug 

> But then I am not sure if there is a use case for this really.

There may be. But let's start with the simple case and add if needed.

Reply:
I gave the library a go but  it needs some fixing, for instance it doesn't handle checkbox properly... I got it to work but at this stage It would probably be easier to fix my solution.

Reply:
Cool. Go ahead with your solution. This was just an idea. If we want to use that library, we should probably do that in a separate pull request anyway. Can we fix the library and submit a PR there, too?

Reply:
We could, although the library hasn't got any test, so I would want to add some before fixing everything. At this point I am not sure it's worth the effort, and we don't know how responsive the maintainer is.
The main advantage I can see is that we could use the library outside stimulus controller.